### ISSUE TITLE
Feature request: inlining

### ISSUE THREAD

**Issue by seanmcl** [souffle-lang/souffle#439]
Currently souffle creates a table for every named predicate.  For example,

```
.decl small_aux_table(x: huge_type).
.decl large_table_with_semantic_meaning(x: huge_type).
small_aux_table(c5).
small_aux_table(c6).
large_table_with_semantic_meaning(X) :- !small_aux_table(X).
...lots of predicates referring to large_table_with_semantic_meaning...
```

Even though the name with semantic content is nicer to read in the body of the program, we currently can't write it because it will be reified, i.e. the large table will be generated.  I propose some way to specify that a predicate should be inlined, e.g.

```
.decl small_aux_table(x: huge_type).
.decl large_table_with_semantic_meaning(x: huge_type) inline.
```

There could be simple syntactic resrictions on inlined rules, such as exactly one definition. While in principle the compiler could try to infer inlining, as in more standard programming languages, in this case I'd just like to ensure that what I know from the outset will be huge doesn't get generated.

Of course, the semantics of `inline` would be capture-avoiding substitution.  

WDYT?

**Comment by scyptnex** [comment]
Automatic inlining optimisations could be related to the much anticipated #23 

In that event, it would be necessary to have a syntax for it so users can adopt manually.

**Comment by b-scholz** [comment]
I believe it is a good idea to support inlining via annotation.  If we have the inline annotation, we need to ensure that we do not have direct recursive rules for the "inlined" relation. For example,

```
.decl P(x:symbol, y:symbol) inline
P(x,y) :- E(x,y).
P(x,z) :- E(x,y), P(y,z). 
```
would not work because there is an occurrence of P in the head and in the body. An alternative approach is to automate inlining, which I feel is hard.

Agreed - for achieving high performance we will need to fully automate query scheduling. This is not the case at the moment. 

Inlining will also influence the magic set transformation.

**Comment by mmcgr** [comment]
Not just #23, but #192 .

**Comment by mmcgr** [comment]
Implemented by #497
